src/parse/lexer.rs

```diff
@@ -0,0 +1,725 @@
+use crate::parse::error::{LexicalError, LexicalErrorType};
+use crate::parse::token::Tok;
+use std::char;
+
+// ideas:
+// * look up 2 / 3 char tokens like keywords
+//   * perf test
+// * track nesting per type and lex error on invalid nesting
+//   * useful?
+
+pub struct Lexer<T: Iterator<Item = (usize, char)>> {
+    chars: T,
+    nesting: usize, // Amount of (), [], or {}
+    pending: Vec<Spanned>,
+    chr0: Option<char>,
+    chr1: Option<char>,
+    loc0: usize,
+    loc1: usize,
+    location: usize,
+}
+pub type Spanned = (usize, Tok, usize);
+pub type LexResult = Result<Spanned, LexicalError>;
+
+pub fn str_to_keyword(word: &str) -> Option<Tok> {
+    // Alphabetical keywords:
+    match word {
+        "as" => Some(Tok::As),
+        "assert" => Some(Tok::Assert),
+        "case" => Some(Tok::Case),
+        "const" => Some(Tok::Const),
+        "external" => Some(Tok::External),
+        "fn" => Some(Tok::Fn),
+        "if" => Some(Tok::If),
+        "import" => Some(Tok::Import),
+        "let" => Some(Tok::Let),
+        "opaque" => Some(Tok::Opaque),
+        "pub" => Some(Tok::Pub),
+        "todo" => Some(Tok::Todo),
+        "try" => Some(Tok::Try),
+        "tuple" => Some(Tok::Tuple),
+        "type" => Some(Tok::Type),
+        _ => None,
+    }
+}
+
+pub fn make_tokenizer<'a>(source: &'a str) -> impl Iterator<Item = LexResult> + 'a {
+    let nlh = NewlineHandler::new(source.char_indices());
+    Lexer::new(nlh)
+}
+
+// The newline handler is an iterator which collapses different newline
+// types into \n always.
+pub struct NewlineHandler<T: Iterator<Item = (usize, char)>> {
+    source: T,
+    chr0: Option<(usize, char)>,
+    chr1: Option<(usize, char)>,
+}
+
+impl<T> NewlineHandler<T>
+where
+    T: Iterator<Item = (usize, char)>,
+{
+    pub fn new(source: T) -> Self {
+        let mut nlh = NewlineHandler {
+            source,
+            chr0: None,
+            chr1: None,
+        };
+        nlh.shift();
+        nlh.shift();
+        nlh
+    }
+
+    fn shift(&mut self) -> Option<(usize, char)> {
+        let result = self.chr0;
+        self.chr0 = self.chr1;
+        self.chr1 = self.source.next();
+        result
+    }
+}
+
+impl<T> Iterator for NewlineHandler<T>
+where
+    T: Iterator<Item = (usize, char)>,
+{
+    type Item = (usize, char);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Collapse \r\n into \n
+        while let Some((i, '\r')) = self.chr0 {
+            if let Some((_, '\n')) = self.chr1 {
+                // Transform windows EOL into \n
+                self.shift();
+            } else {
+                // Transform MAC EOL into \n
+                self.chr0 = Some((i, '\n'))
+            }
+        }
+
+        self.shift()
+    }
+}
+
+impl<T> Lexer<T>
+where
+    T: Iterator<Item = (usize, char)>,
+{
+    pub fn new(input: T) -> Self {
+        let mut lxr = Lexer {
+            chars: input,
+            nesting: 0,
```

Comment:
Why is it that nesting is tracked in the lexer rather than the parser?

Reply:
My idea was that since in the grammar you must always close inner before outer nestings, it would be easier to point out the error precisely in the lexer. See this response for a little more detail on use: https://github.com/gleam-lang/gleam/pull/891#discussion_r549553214

It could be done in the Parser, you could scan tokens as they come out of the lexer in `next_tok` and accomplish the same thing. I think my mind felt it was easier at the lexing phase. Probably just a working thought thing.